### ISSUE TITLE
bleach 2.1.3 raises KeyError on unknown entity in URI

### ISSUE THREAD

**Issue by jwhitlock** [mozilla/bleach#359]
```
>>> import bleach
>>> bleach.clean('<a href="&xx;https://foo.com">')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/bleach/__init__.py", line 99, in clean
    return cleaner.clean(text)
  File "/usr/local/lib/python2.7/site-packages/bleach/sanitizer.py", line 282, in clean
    return self.serializer.render(filtered)
  File "/usr/local/lib/python2.7/site-packages/html5lib/serializer.py", line 398, in render
    return "".join(list(self.serialize(treewalker)))
  File "/usr/local/lib/python2.7/site-packages/bleach/sanitizer.py", line 751, in serialize
    for stoken in super(BleachHTMLSerializer, self).serialize(treewalker, encoding):
  File "/usr/local/lib/python2.7/site-packages/html5lib/serializer.py", line 265, in serialize
    for token in treewalker:
  File "/usr/local/lib/python2.7/site-packages/bleach/sanitizer.py", line 439, in __iter__
    ret = self.sanitize_token(token)
  File "/usr/local/lib/python2.7/site-packages/bleach/sanitizer.py", line 469, in sanitize_token
    return self.allow_token(token)
  File "/usr/local/lib/python2.7/site-packages/bleach/sanitizer.py", line 620, in allow_token
    new_value = self.sanitize_uri_value(val, self.allowed_protocols)
  File "/usr/local/lib/python2.7/site-packages/bleach/sanitizer.py", line 557, in sanitize_uri_value
    new_value = convert_entities(value)
  File "/usr/local/lib/python2.7/site-packages/bleach/sanitizer.py", line 123, in convert_entities
    new_text.append(convert_entity(entity))
  File "/usr/local/lib/python2.7/site-packages/bleach/sanitizer.py", line 101, in convert_entity
    return ENTITIES[value]
KeyError: u'xx'
```

Found by a developer.mozilla.org user trying out some XSS vectors.

**Comment by willkg** [comment]
Oh, hrm... I'll try to fix this next week and push out an update.

**Comment by jwhitlock** [comment]
Not a rush (just breaking the researcher’s experience) and I can submit a PR if it would help.

I couldn’t find any guidance in the standards for how to process an invalid entity. My first thought is to remove it - replace with an empty string.

**Comment by willkg** [comment]
This is an ambiguous ampersand:

* https://html.spec.whatwg.org/multipage/syntax.html#syntax-ambiguous-ampersand
* https://html.spec.whatwg.org/multipage/parsing.html#ambiguous-ampersand-state

The parsing spec says that a parser should not resolve it:

* https://html.spec.whatwg.org/multipage/parsing.html#parse-error-unknown-named-character-reference

I think that means we need to keep it in the text. Thus we should end up with something like this:

```
>>> import bleach
>>> bleach.clean('<a href="&xx;https://foo.com">')
'<a href="&xx;https://foo.com"></a>'
```

@gsnedders Does that sound right to you? ^^^

**Comment by gsnedders** [comment]
@willkg `<a href="&xx;https://foo.com"></a>` or `<a href="&amp;xx;https://foo.com"></a>` are both acceptable; I'd lean towards the latter.

**Comment by willkg** [comment]
@gsnedders Thank you!

**Comment by andriykohut** [comment]
👋 Doesn't seem like this got into `'2.1.3'` release. I can't replicate this with a freshly installed package but it works if I install bleach from the `master` branch.

**Comment by willkg** [comment]
@andriykohut 2.1.3 was released on March 5th and this was fixed on March 19th, so it wasn't done for 2.1.3. Hope that helps!

**Comment by willkg** [comment]
I just released 2.1.4 that includes this fix. https://pypi.org/project/bleach/2.1.4/
